Issue solutions

Vlamboljant opened the pull request:
Possible solutions for:
     Issue #576 
     Issue #577 
     Issue #571 
     Issue #587

mattbenic commented on Script/AtomicEditor/ui/modal/build/BuildSettingsWindow.ts:
```diff
@@ -41,7 +41,7 @@ export class BuildSettingsWindow extends ModalWindow {
         platformSource.addItem(new Atomic.UISelectItem("Mac", "MacBuildSettings", "LogoMac"));
         platformSource.addItem(new Atomic.UISelectItem("Android", "AndroidBuildSettings", "LogoAndroid"));
-        platformSource.addItem(new Atomic.UISelectItem("iOS", "iOSBuildSettings", "LogoIOS"));
+        platformSource.addItem(new Atomic.UISelectItem("iOS", "iosBuildSettings", "LogoIOS"));
```
Wouldn't this be iOsBuildSettings?

JoshEngebretson replied:
iOS is kind of a wildcard, liking iosBuildSettings in terms of a "rule" I guess, could go either way, let's stick with this...

JoshEngebretson replied:
I made a comment above, I think consistent case for iOS, so:

iosBuildSettings and licenseIOS read the best...

JoshEngebretson commented:
Great, a bunch of small issues in a single pull :)  Can we squish the commit history on this down?  I am trying to get better about this myself.  

The changes in case here could also potentially affect examples.  I have to look through them.  The case problem is a good thing to clean up at this point, thankfully we have static typing on the editor scripts!

Also, good job on weaving between native code and script.  Let's get this passing CI and then I'll follow up with any notes.  THANKS!

@AtomicBuildBot ok to test

mattbenic commented:
Hi Josh, will squashing be safe after your build CI has already pulled down these commits? Or does it just do a clean pull so it doesn't matter?

BTW What does your build bot do?

JoshEngebretson commented:
@Vlamboljant This is looking great, I may still have a note or two, though only landing minor fixes to master for the next few days, so will be a little bit before landing it.

@mattbenic That is a good point, let's hold off on the squash until we can try one through CI and I am in position to update the build scripts.  

The AtomicBuildBot is a Jenkins thing (which is free and rocks, used by Microsoft CoreCLR team, btw), like if I:

@AtomicBuildBot add to whitelist

This adds @Vlamboljant to the whitelist for PR's to go straight to CI, which is now done :)  Otherwise, people can spam your build servers... pretty important when everything on public repos 

- Josh

JoshEngebretson commented on Script/AtomicEditor/ui/frames/menus/MainFrameMenu.ts:
```diff
@@ -201,6 +202,14 @@ class MainFrameMenu extends Atomic.ScriptObject {
+                myPrefs.useDefaultConfig();
+                myPrefs.saveEditorWindowData(myPrefs.editorWindow);
+                myPrefs.savePlayerWindowData(myPrefs.playerWindow);
+                Atomic.getEngine().exit();
```
Nice!

JoshEngebretson commented on Script/AtomicEditor/ui/modal/About.ts:
```diff
@@ -90,7 +90,7 @@ class About extends ModalWindow {
             availableText += "<widget TBSkinImage: skin: 'LogoAndroid-Small'> ";
 
-            if (licenseSystem.licenseIOS)
+            if (licenseSystem.licenseIos)
```
Man, iOS is so gnarly, every CamelCase always looks wrong with it.  This probably should be licenseIOS though, licenseIos looks pretty weird and is difficult to "see".

JoshEngebretson commented on Script/AtomicEditor/ui/modal/build/platforms/AndroidSettingsWidget.ts:
```diff
@@ -101,8 +101,8 @@ class AndroidSettingsWidget extends Atomic.UIWidget implements BuildSettingsWind
-                    if (toolPrefs.jDKRootPath != path) {
-                        toolPrefs.jDKRootPath = path;
+                    if (toolPrefs.jdkRootPath != path) {
+                        toolPrefs.jdkRootPath = path;
```
Much better!

JoshEngebretson commented on Script/AtomicEditor/ui/modal/build/platforms/AndroidSettingsWidget.ts:
```diff
@@ -181,7 +181,7 @@ class AndroidSettingsWidget extends Atomic.UIWidget implements BuildSettingsWind
             ToolCore.toolEnvironment.toolPrefs.antPath = this.antPathEdit.text;
 
-        this.settings.sDKVersion = this.sdkTargetSelect.text;
+        this.settings.sdkVersion = this.sdkTargetSelect.text;
```
Man, I remember having to type this and thinking wish I had time to fix it... that time is now, thanks!

JoshEngebretson commented on Script/AtomicEditor/ui/playmode/PlayerOutput.ts:
```diff
@@ -47,7 +47,8 @@ class PlayerOutput extends Atomic.UIWindow {
             this.output.text = "";
 
-        this.output.appendText(ev.message + "\n");
+        this.output.appendText("\n" + ev.message);//Ensures all subsequent \n operators and player text outputs correctly
```
Let's put comments directly above the line they are commenting.  We'll have linting soon which should catch stuff like this...

JoshEngebretson commented on Source/AtomicEditor/Application/AEEditorApp.cpp:
```diff
@@ -76,6 +76,11 @@ void AEEditorApp::Setup()
 
+    // Ensure exclusive fullscreen is disabled in Editor application
+    Input* input = GetSubsystem<Input>();
+    input->SetToggleFullscreen(false);
```
Joy!

JoshEngebretson commented:
Ok, sorry it took awhile to review this, great work! I think we just need to address the iOS case handling, get you added to the contributors (https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/CONTRIBUTING.md), and can land!

mattbenic commented:
I think the issue with iOS was that the code in c++ added to correct case for js bindings will always convert iOS to iOs. Perhaps it needs a special test just for that case (and others like it that pop up) to keep the case you want. It's nasty but not unreasonable.

JoshEngebretson commented:
Yes, even if it checks specific iOS string combinations (licenseIOS or iosBuildSettings for example) to "correct" case, this would be fine.

Vlamboljant commented:
Issue #654 I changed how the IOS is treated as specified in the notes. Also changed the commenting style in PlayerOutput.ts. 
Issue #437 We couldn't find or reproduce the case that was causing this problem, but we did handle the case of a private constructor and if a new case arises that it can be handled in the same way.

JoshEngebretson commented:
Thanks!  A few notes on the latest changes :)

Vlamboljant commented:
Changed JSBClass to be checked if it has a private constructor. Changed JSBFunction to check if function is a private constructor and modified the JSBHeaderVisitor to not add functions that are not public, but marks the class if it has a private constructor.

JoshEngebretson commented:
It might be a good idea to do 1 PR per issue for now.  It is getting a little difficult to track what is going on and we're holding up issues while adding new issues to this PR during code review.  THANKS! :)

mattbenic commented:
Yup, it's building up to a crazy level. Will have them branch per issue in future

JoshEngebretson commented:
No worries, the help is really appreciated.  Maybe it would be good to back the #437 changes out of this PR and finish up the review on the other stuff, which was close to landing?

mattbenic commented:
I'm splitting these into specific branches per issue, and will PR those separately. @JoshEngebretson can you cancel this PR?

JoshEngebretson commented:
@mattbenic Thanks, closing this PR
